File: python/taflow/average_price.py
"""Canonical average-price adapter."""

from ._native import StatefulAvgprice
from ._price_state import OhlcPriceState


class AveragePrice(OhlcPriceState):
    """Compute average price through the native Rust kernel

    Parameters
    ----------
    Input series and configuration values are accepted by the constructor.

    Returns
    -------
    AveragePrice
        A persistent native-backed indicator adapter.
    """

    _native_cls = StatefulAvgprice

    def append(self, _open: object, high: object, low: object, close: object) -> "AveragePrice":
        """Append one observation and return this indicator."""
        super().append(_open, high, low, close)
        return self

    def extend(self, _open: object, high: object, low: object, close: object) -> "AveragePrice":
        """Append aligned histories and return this indicator."""
        super().extend(_open, high, low, close)
        return self

    def reset(self) -> "AveragePrice":
        """Reset native state and return this indicator."""
        super().reset()
        return self

File: crates/taflow-core/src/stream/average_price.rs
//! Batch implementation for `average_price`.

use super::price_transform::*;
use crate::error::{TaError, TaResult};

/// Compute the average price result for the supplied aligned series.
///
/// # Parameters
///
/// * `open` - Input series or configuration value.
/// * `high` - Input series or configuration value.
/// * `low` - Input series or configuration value.
/// * `close` - Input series or configuration value.
///
/// # Returns
///
/// An aligned result with TA-Lib-compatible validation and warm-up values.
pub fn average_price(open: &[f64], high: &[f64], low: &[f64], close: &[f64]) -> TaResult<Vec<f64>> {
    let len = open.len();
    validate_ohlc_len(len, high, low, close)?;
    let mut output = Vec::with_capacity(len);
    output.extend(
        open.iter()
            .zip(high.iter())
            .zip(low.iter())
            .zip(close.iter())
            .map(|(((&o, &h), &l), &c)| (o + h + l + c) * 0.25),
    );
    Ok(output)
}
